### PR TITLE
feat: lofn integration

### PR DESCRIPTION
**Pull request by capJavert**

### CODE REVIEW

File: src/integrations/feed/configs.ts [hunk 1]
```diff
@@ -111,3 +121,43 @@ export class FeedUserStateConfigGenerator implements FeedConfigGenerator {
...
+    this.baseConfig = baseConfig;
+    this.lofnClient = new LofnClient();
+    this.opts = opts;
+  }
+
+  async generate(
+    ctx: Context,
+    opts: DynamicConfig,
+  ): Promise<FeedConfigGeneratorResult> {
+    return runInSpan('FeedLofnConfigGenerator', async () => {
```
Comment (idoshamun) [inline src/integrations/feed/configs.ts:145]: If it ever becomes a bottleneck we can run `fetchConfig` and `FeedPreferencesConfigGenerator` in parallel. Probably will require changing a little bit `FeedPreferencesConfigGenerator`. We can consider it as part of the cleanup

File: src/integrations/feed/generators.ts [hunk 2]
```diff
@@ -151,5 +157,17 @@ export const feedGenerators: Record<FeedVersion, FeedGenerator> = Object.freeze(
...
+  if (version >= 30) {
+    return new FeedGenerator(
+      feedClient,
+      new FeedLofnConfigGenerator(
+        {},
+        {
+          feed_version: version.toString() as FeedVersion,
+        },
+      ),
+    );
```
Comment (idoshamun) [inline src/integrations/feed/generators.ts:169]: During the cleanup, I would rethink it a little bit

File: src/integrations/feed/types.ts [hunk 3]
```diff
@@ -57,8 +58,16 @@ export type FeedConfig = {
 
 export type DynamicConfig = Omit<FeedConfig, 'total_pages'>;
 
+export type FeedConfigGeneratorResult = {
+  config: FeedConfig;
+  tyr_metadata?: TyrMetadata;
```
Comment (idoshamun) [inline src/integrations/feed/types.ts:63]: Check my comment about making it more generic

File: src/integrations/feed/clients.ts [hunk 4]
```diff
@@ -23,7 +23,12 @@ export class FeedClient implements IFeedClient {
     this.fetchOptions = fetchOptions;
   }
 
-  async fetchFeed(ctx, feedId, config): Promise<FeedResponse> {
+  async fetchFeed(
+    ctx,
+    feedId,
+    config,
+    tyr_metadata = undefined,
```
[the remark this answers is not in the document]
Comment (capJavert) [reply src/integrations/feed/clients.ts]: Yes, its just a variable name, I will rename to `extraMetadata`, good point.

File: src/integrations/feed/configs.ts (hunk 1 above)
Comment (capJavert) [reply src/integrations/feed/configs.ts:142]: Yes, agreed 👍

File: src/integrations/feed/generators.ts (hunk 2 above)
Comment (capJavert) [reply src/integrations/feed/generators.ts]: Yes, basically the issue was that FeedConfig and feed_config_name gets passed around for other generators and I did not want to add `feed_version` to FeedConfig type. Will note it.

File: src/integrations/feed/types.ts (hunk 3 above)
Comment (capJavert) [reply src/integrations/feed/types.ts:63]: Yes!

File: src/integrations/feed/configs.ts (hunk 1 above)
Comment (capJavert) [reply src/integrations/feed/configs.ts:142]: I'll take a look, it should not be too bad to implement it in this PR.

File: src/integrations/feed/clients.ts [hunk 5]
```diff
@@ -37,10 +43,19 @@ export class FeedClient implements IFeedClient {
...
-      ]),
+      data: res.data.map(({ post_id, metadata }) => {
+        const hasMetadata = !!(metadata || extraMetadata);
+
+        return [
+          post_id,
+          (hasMetadata &&
+            JSON.stringify({
+              ...metadata,
+              mab: extraMetadata,
```
Comment (idoshamun) [inline src/integrations/feed/clients.ts:54]: Originally this is what I meant
```suggestion
              ...extraMetadata,
```

Comment (capJavert) [reply src/integrations/feed/clients.ts:54]: Ahh yes, will update.

File: src/integrations/feed/configs.ts (hunk 1 above)
Comment (capJavert) [reply src/integrations/feed/configs.ts:147]: Done 82511f866aaf8b1c58df9a9d1fb977e342e20b09

File: src/integrations/feed/configs.ts [hunk 6]
```diff
@@ -111,3 +121,52 @@ export class FeedUserStateConfigGenerator implements FeedConfigGenerator {
...
+      const [lofnConfig, preferencesConfig] = await Promise.all([
+        this.lofnClient.fetchConfig({
+          user_id: opts.user_id,
+          feed_version: this.opts.feed_version,
+        }),
+        this.feedPreferencesConfigGenerator.generate(ctx, opts),
```
Comment (rebelchris) [inline src/integrations/feed/configs.ts:153]: What would this fetch?
Isn't it just one call to lofn to include everything?

File: src/integrations/lofn/clients.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  constructor(
+    url = process.env.LOFN_ORIGIN,
+    fetchOptions: RequestInit = globalFetchOptions,
+  ) {
+    this.url = url;
+    this.fetchOptions = fetchOptions;
+  }
+
+  fetchConfig(payload: LofnFeedConfigPayload): Promise<LofnFeedConfigResponse> {
+    return retryFetchParse(
```
Comment (rebelchris) [inline src/integrations/lofn/clients.ts:23]: Optional, since it's new maybe we'd like specific logging for fails?

**Review by rebelchris — APPROVED** [review]
Nothing blocking from my side. more questions :)

File: src/integrations/feed/configs.ts (hunk 6 above)
Comment (rebelchris) [reply src/integrations/feed/configs.ts:153]: Oh got you! 🙏

File: src/integrations/feed/configs.ts [hunk 8]
```diff
@@ -129,31 +129,43 @@ type FeedLofnConfigGeneratorOptions = {
...
+      ]);
+
+      const config = getDefaultConfig(
+        {
+          ...preferencesConfig.config,
+          total_pages: lofnConfig.config.total_pages,
```
Comment (idoshamun) [inline src/integrations/feed/configs.ts:159]: We don't really need to take lofn total pages. We can keep it owned by API just like all other pagination params

**Review by idoshamun — APPROVED** [review]
Just apply my comment, but looks amazing

File: src/integrations/lofn/clients.ts (hunk 7 above)
Comment (capJavert) [reply src/integrations/lofn/clients.ts:23]: Added in 765337e984221288565f15363bbde8457940b2c7

File: src/integrations/feed/configs.ts (hunk 8 above)
Comment (capJavert) [reply src/integrations/feed/configs.ts]: `page_size` as well?

Comment (capJavert) [reply src/integrations/feed/configs.ts]: These should take from API?
- fresh_page_size
- page_size
- total_pages

Comment (idoshamun) [reply src/integrations/feed/configs.ts]: yes

**Comment by idoshamun** [comment]
@capJavert can you please log the final configuration that are being generated by the lofn config generator? it would be useful while we debug. Include also the feed version and user id

**Comment by capJavert** [comment]
@idoshamun added logging.

**Pull request merged**